### PR TITLE
Fallback to vaCreateImage + vaPutImage when vaDeriveImage fails

### PR DESCRIPTION
**Pull request by edman007**
Per intel docs for libva, when vaDeriveImage fails vaCreateImage +
  vaPutImage should be tried. This is important as mesa with AMD HW
  will always fail because the image is interlaced, but the HW
  capability still exists

Fixes https://github.com/opencv/opencv/issues/21536

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ x ] I agree to contribute to the project under Apache 2 License.
- [ x ] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [ x ] The PR is proposed to the proper branch
- [ x ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ x ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by edman007** [comment]
Fixed looping through the array of formats to use the num_formats returned by libva, and I have tested this on mine, works just fine

File: modules/core/src/va_intel.cpp [hunk 1]
```diff
@@ -608,8 +608,32 @@ void convertToVASurface(VADisplay display, InputArray src, VASurfaceID surface,
...
+                CV_Error(cv::Error::StsError, "VA-API: vaQueryImageFormats failed");
+            VAImageFormat *selected_format = &fmt_list[0];
+            for (int i = 0; i < num_formats; i++){
+                if (fmt_list[i].fourcc == VA_FOURCC_NV12 || fmt_list[i].fourcc == VA_FOURCC_YV12){
+                    selected_format = &fmt_list[i];
+                    break;
+                }
+            }
+
+            status = vaCreateImage(display, selected_format, size.width, size.height, &image);
```
Comment (alalek) [inline modules/core/src/va_intel.cpp:627]: What if `selected_format` is not "found"? Why do we fallback to the first one?

I believe we should emit error message as fast as possible.

**Review by alalek — COMMENTED** [review]
Thank you for contribution!

`convertToVASurface()` workaround doesn't look valid.
We need to update data of the surface. Updating image is not enough as it doesn't update content of the original surface.

**Comment by edman007** [comment]
@alalek Updated it with the changes you suggested

- Throw an error if num_formats <= 0
- Use an std::vector for the formats
- Throw an error if none of the supported formats are found

> convertToVASurface() workaround doesn't look valid.
> We need to update data of the surface. Updating image is not enough as it doesn't update content of the original surface.

That should be fixed now! It now uses vaPutImage or vaGetImage, and vaPutImage is after we write the data to the image. I have no idea why vaPutImage worked for me in my testing when it should have been vaGetImage, but I tested it in both directions and verified that I can modify it as a UMat and write it back and encode it with VA-API/FFmpeg

**Review by alalek — APPROVED** [review]
Thank you for contribution :+1:

**Pull request merged**